neqo-transport/src/cc.rs

```diff
@@ -118,19 +118,17 @@ impl CongestionControl {
         let last_lost_pkt = lost_packets.last().unwrap();
         self.on_congestion_event(now, last_lost_pkt.time_sent);
 
-        let in_persistent_congestion = {
-            let congestion_period = pto * PERSISTENT_CONG_THRESH;
+        let congestion_period = pto * PERSISTENT_CONG_THRESH;
 
-            match largest_acked_sent {
-                Some(las) => las < last_lost_pkt.time_sent - congestion_period,
-                None => {
-                    // Nothing has ever been acked. Could still be PC.
-                    let first_lost_pkt_sent = lost_packets.first().unwrap().time_sent;
-                    last_lost_pkt.time_sent - first_lost_pkt_sent > congestion_period
-                }
-            }
-        };
-        if in_persistent_congestion {
+        let first_lost_pkt_after_largest_acked = lost_packets
+            .iter()
+            .find(|pkt| Some(pkt.time_sent) > prev_largest_acked_sent);
+
+        let starting_time_sent = first_lost_pkt_after_largest_acked
```

Comment:
"The congestion period is calculated as the time between the oldest and newest lost packets"

This code sets this to `first_lost_pkt_after_largest_acked` first.  You could instead say `lost_packets.last().unwrap()`, but that doesn't work either because if some packet in between is acknowledged, then you aren't in persistent congestion as it is defined as "loss of all in-flight packets sent over a long enough period of time".  So what we need is something trickier.

```rust
let mut start = None;
let mut end = now;
let mut last_pn = 0;
let mut iter = lost_packets.iter().take_while(|p| p.time_sent <= prev_largest_acked_sent);
if let Some(p) = iter.next() {
    let mut start = p.time_sent;
    for p in iter {
        // Not a contiguous range of lost packets, start over.
        if p.pn != last_pn + 1 {
            start = p.time_sent;
        } else if p.time_sent.duration_since(start) > congestion_period {
            // In persistent congestion.  Stop.
            self.congestion_window = MIN_CONG_WINDOW;
            break;
        }
        last_pn = p.pn; // Note that we would need to start tracking packet number in SentPacket this for this check to work.
    }
}
```